lib/fog/core/cache.rb

```diff
@@ -0,0 +1,183 @@
+  #
+  #   instances = Fog::Cache.load(Fog::Compute::AWS::Server, compute)
+  #   instances.first.id # => "i-0569a70ae6f47d229"
+  #
+  # Note that if there is no cache located for the +model+ class and +service+ passed to `Fog::Cache.load`
+  # you will get an exception you can handle (for example, to load the resources for the fisrt time):
+  #
+  #   Fog::Cache.expire_cache!(Fog::Compute::AWS::SecurityGroup, compute)
+  #   # ... now there is no SecurityGroup cache data. So, if you tried to load it, you would get an exception:
+  #
+  #   Fog::Cache.load(Fog::Compute::AWS::SecurityGroup, computek)
+  #     rescue Fog::Cache::CacheNotFound => e
+  #       puts "could not find any cache data for security groups on #{compute}"
+  #       get_resources_and_dump
+  #
+  # == Extending cache backends
+  #
+  # Currently this is on-disk using yml. If need be, this could be extended to other cache backends:
+  #
+  # Find references of yaml in this file, split out to strategy objects/diff backends etc.
+  class Cache
+    class CacheNotFound < StandardError; end
+
+    # when a resource is used such as `server.cache.dump` the model klass is passed in
+    # so that it can be identified from a different session.
+    attr_reader :model
+
+    # Loads cache associated to the +model_klass+ and +service+ into memory.
+    #
+    # If no cache is found, it will raise an error for handling:
+    #
+    #   rescue Fog::Cache::CacheNotFound
+    #     set_initial_cache
+    #
+    def self.load(model_klass, service)
+      cache_files = Dir.glob("#{sandbox(model_klass, service)}/*")
+
+      raise CacheNotFound if cache_files.empty?
+
+      loaded = cache_files.map do |path|
+        model_klass = Object.const_get(cache_data(path)[:model_klass])
+        model_klass.new(cache_data(path)[:attrs])
+      end
+
+      collection_klass = cache_data(cache_files.sample)[:collection_klass] &&
+                         Object.const_get(cache_data(cache_files.sample)[:collection_klass])
+
+      loaded.each do |i|
+        # See https://github.com/fog/fog-aws/issues/354#issuecomment-286789702
+        i.collection = collection_klass.new(:service => service) if collection_klass
+        i.instance_variable_set(:@service, service)
+      end
+    end
+
+    # Expires cache - this does not expire all cache associated.
+    # Instead, this will remove all on-disk cache of this specific +model_klass+ and and +@service+
+    def self.expire_cache!(model_klass, service)
+      FileUtils.rm_rf(sandbox(model_klass, service))
+    end
+
+    # loads yml cache from path on disk
+    def self.cache_data(path)
+      @cache_data ||= {}
+      return @cache_data[path] if @cache_data[path]
+      @cache_data[path] = YAML.load(File.read(path))
+    end
+
+    # The path/namespace where the cache is stored for a specific +model_klass+ and +@service+.
+    def self.sandbox(model_klass, service)
```

Comment:
Same thing here. I would expect this method to create a sandbox or something like that instead of returning a name.